File: LeetCode/1254. Number of Closed Islands/solution.cpp
/**
 * 47 / 47 test cases passed.
 * Runtime: 12 ms
 * Memory Usage: 9.1 MB 
 */
class Solution {
public:
    int direct[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    int m, n;
    void dfs(vector<vector<int>> &grid, int x, int y) {
        grid[x][y] = 1;
        for (auto &d : direct) {
            int xx = x + d[0];
            int yy = y + d[1];
            if (xx < 0 || xx >= m || yy < 0 || yy >= n) continue;
            if (grid[xx][yy]) continue;
            dfs(grid, xx, yy);
        }
    }
    int closedIsland(vector<vector<int>>& grid) {
        m = grid.size(), n = grid[0].size();
        for (int i = 0; i < m; ++ i) {
            if (!grid[i][0]) dfs(grid, i, 0);
            if (!grid[i][n - 1]) dfs(grid, i, n - 1);
        }
        for (int i = 1; i < n - 1; ++ i) {
            if (!grid[0][i]) dfs(grid, 0, i);
            if (!grid[m - 1][i]) dfs(grid, m - 1, i);
        }
        int ans = 0;
        for (int i = 1; i < m - 1; ++ i) {
            for (int j = 1; j < n - 1; ++ j) {
                if (!grid[i][j]) {
                    ++ ans;
                    dfs(grid, i, j);
                }
            }
        }
        return ans;
    }
};

File: LeetCode/1254. Number of Closed Islands/solution.py
"""
47 / 47 test cases passed.
Runtime: 64 ms
Memory Usage: 15.5 MB
"""
class Solution:
    def closedIsland(self, grid: List[List[int]]) -> int:
        ans, m, n = 0, len(grid), len(grid[0])
        def dfs(x, y):
            grid[x][y] = 1
            for xx, yy in [(x + 1, y), (x - 1, y), (x, y + 1), (x, y - 1)]:
                if 0 <= xx < m and 0 <= yy < n and grid[xx][yy] == 0:
                    dfs(xx, yy)
        
        for i in range(m):
            if not grid[i][0]: dfs(i, 0)
            if not grid[i][n - 1]: dfs(i, n - 1)
        for i in range(1, n - 1):
            if not grid[0][i]: dfs(0, i)
            if not grid[m - 1][i]: dfs(m - 1, i)
        for i in range(1, m - 1):
            for j in range(1, n - 1):
                if not grid[i][j]:
                    ans += 1
                    dfs(i, j)
        return ans
